gio/src/subclass/initable.rs

```diff
@@ -0,0 +1,177 @@
+
+use crate::Cancellable;
+use crate::Initable;
+
+pub trait InitableImpl: ObjectImpl + Send {
+    fn init(&self, initable: &Self::Type, cancellable: Option<&Cancellable>) -> Result<(), Error>;
+}
+
+pub trait InitableImplExt: ObjectSubclass {
+    fn parent_init(
+        &self,
+        initable: &Self::Type,
+        cancellable: Option<&Cancellable>,
+    ) -> Result<(), Error>;
+}
+
+impl<T: InitableImpl> InitableImplExt for T {
+    fn parent_init(
+        &self,
+        initable: &Self::Type,
+        cancellable: Option<&Cancellable>,
+    ) -> Result<(), Error> {
+        unsafe {
+            let type_data = Self::type_data();
+            let parent_iface =
+                type_data.as_ref().parent_interface::<Initable>() as *const ffi::GInitableIface;
+
+            let func = (*parent_iface)
+                .init
+                .expect("no parent \"init\" implementation");
+
+            let mut err = ptr::null_mut();
+            func(
+                initable.unsafe_cast_ref::<Initable>().to_glib_none().0,
+                cancellable.to_glib_none().0,
+                &mut err,
+            );
+
+            if err.is_null() {
+                Ok(())
+            } else {
+                Err(from_glib_full(err))
+            }
+        }
+    }
+}
+
+unsafe impl<T: InitableImpl> IsImplementable<T> for Initable {
+    fn interface_init(iface: &mut glib::Interface<Self>) {
+        let iface = iface.as_mut();
+        iface.init = Some(initable_init::<T>);
+    }
+
+    fn instance_init(_instance: &mut glib::subclass::InitializingObject<T>) {}
+}
+
+unsafe extern "C" fn initable_init<T: InitableImpl>(
+    initable: *mut ffi::GInitable,
+    cancellable: *mut ffi::GCancellable,
+    error: *mut *mut glib::ffi::GError,
+) -> glib::ffi::gboolean {
+    let instance = &*(initable as *mut T::Instance);
+    let imp = instance.impl_();
+
+    match imp.init(
+        from_glib_borrow::<_, Initable>(initable).unsafe_cast_ref(),
+        Option::<Cancellable>::from_glib_borrow(cancellable)
+            .as_ref()
+            .as_ref(),
+    ) {
+        Ok(()) => glib::ffi::GTRUE,
+        Err(e) => {
+            if !error.is_null() {
+                *error = e.into_raw();
+            }
+            glib::ffi::GFALSE
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use glib::object::*;
+
+    pub mod imp {
+        use super::*;
+        use crate::Cancellable;
+        use crate::Initable;
+        use std::cell::Cell;
+
+        pub struct InitableTestType(pub Cell<u64>);
+
+        #[glib::object_subclass]
+        impl ObjectSubclass for InitableTestType {
+            const NAME: &'static str = "InitableTestType";
+            type Type = super::InitableTestType;
+            type ParentType = glib::Object;
+            type Interfaces = (Initable,);
+
+            fn new() -> Self {
+                Self(Cell::new(0))
+            }
+        }
+
+        impl InitableImpl for InitableTestType {
```

Comment:
Maybe this is a silly question, but what happens when one puts `Initable` in the implemented interfaces in `ObjectSubclass`:
```
type Interfaces = (Initable,);
```
but does not  actually implement `InitableImpl`?

Reply:
Ok, I guess the object will not be`IsImplementable` for the `Initable` interface - probably will fail at compile time?

Reply:
Yes:

```
error[E0277]: the trait bound `subclass::initable::tests::imp::InitableTestType: InitableImpl` is not satisfied
   --> gio/src/subclass/initable.rs:108:13
    |
108 |             type Interfaces = (Initable,);
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `InitableImpl` is not implemented for `subclass::initable::tests::imp::InitableTestType`
```